fp16 support for FullyConnected op

slayton58 opened the pull request:
Includes math lib support, removal of double-precision.

asaadaldien commented on caffe2/operators/utility_ops_gpu.cc:
```diff
@@ -5,6 +5,28 @@
+    return DoRunWithType<float16,float16>();
+  } else {
+    CAFFE_THROW("Unsupported inputs");
+  }
```
return false;

slayton58 commented:
rebased + return false changes made

slayton58 commented on caffe2/utils/conversions.h:
```diff
@@ -0,0 +1,187 @@
+namespace convert {
+
+namespace {
+inline float16 cpu_float2half_rn(float f)
```
Fixed just now - I rebased just before your comment, so I hope that's still good :)

slayton58 commented on caffe2/operators/elementwise_op.cu:
```diff
@@ -222,6 +220,56 @@ bool SumReduceLikeOp<CUDAContext>::DoRunWithType() {
+template <typename T, typename M>
+__global__
+void binary_add_kernel(const int N, const T* a, const T* b, T* r) {
+  for (int idx=threadIdx.x+blockIdx.x*blockDim.x; idx < N; idx+=blockDim.x*gridDim.x) {
```
Missed this before - fixed

asaadaldien commented on caffe2/operators/fully_connected_op_gpu.cc:
```diff
@@ -2,9 +2,51 @@
+                         float16,    // Y
+                         float>(); // Math
+  }
+}
```
else {
CAFFE_THROW("Unsupported type");
return false;
}

slayton58 replied:
I assume the nervana change was meant to be caffe2/contrib/nervana/nervana_math_gpu.cc

Changes made.

Yangqing commented on caffe2/utils/conversions.h:
```diff
@@ -0,0 +1,185 @@
+inline float16 cpu_float2half_rn(float f) {
+    float16 ret;
+
+    unsigned x = *((int*)(void*)(&f));
```
Just for safety... could you do

    static_assert(sizeof(unsigned int) == sizeof(float),
        "Programming error: sizeof(unsigned int) != sizeof(float).");
    unsigned x = *(reinterpret_cast<unsigned int*>(&f));

asaadaldien replied:
Unfortunately both `*((int*)(void*)(&f))` and *(reinterpret_cast<unsigned int*>(&f)) will break builds with -fstrict-aliasing (error: dereferencing type-punned pointer will break strict-aliasing rules)
To fix that we can just use the pointer itself: 
`unsigned x* = reinterpret_cast<unsigned int*>(&f);`
 and in line 101 as well.

slayton58 replied:
I don't understand how this is meant to work if I can't dereference that pointer -- I need the actual value of x as originally done.

asaadaldien replied:
@slayton58 You can dereference the pointer x later but the compiler with -fstrict-aliasing
is not happy with this `*(reinterpret_cast<unsigned int*>(&f))`

slayton58 replied:
ok, so I can do `unsigned *xp = reinterpret_cast<unsigned int*>(&f); unsigned x = *xp;`

asaadaldien replied:
Yes that should work. You can do thins or dereference x in line 33.

slayton58 commented:
Updated: Note that I've reverted the dgemv commit, if we need to support double precision (I was under the impression we no longer did) then there's more work to be done in terms of function signatures.

dragonxlwang commented:
Hey, am wondering what's the reason behind the decision to remove double-precision?

slayton58 commented:
@dragonxlwang It complicates the API and no-one uses double precision.
